### PR TITLE
Add to the info table db_completed_at

### PR DESCRIPTION
**Pull request by anvial**
This PR adds db_completed_at to upgrade_info table to indicate for watching controllers in the cluster when upgrade owner controller finishes the DB upgrade process.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
go test github.com/juju/juju/domain/upgrade/... -gocheck.v
```

## Links

**Jira card:** [JUJU-[4669]](https://warthogs.atlassian.net/browse/JUJU-4669)

### CODE REVIEW

File: domain/schema/controller.go [hunk 1]
```diff
@@ -436,7 +436,8 @@ CREATE TABLE upgrade_info (
     target_version   TEXT NOT NULL,
     created_at       TIMESTAMP NOT NULL,
     started_at       TIMESTAMP,
-    completed_at     TIMESTAMP
+    completed_at     TIMESTAMP,
+    db_completed_at  TIMESTAMP
```
Comment (manadart) [inline domain/schema/controller.go:440]: Move this above `completed at`.

File: domain/upgrade/state/state.go [hunk 2]
```diff
@@ -299,3 +299,38 @@ func (st *State) ActiveUpgrade(ctx context.Context) (string, error) {
...
+		return errors.Trace(err)
+	}
+
+	getUpgradeDBCompletedQuery := "SELECT db_completed_at AS &Info.* FROM upgrade_info WHERE uuid = $M.info_uuid"
+	getUpgradeDBCompletedStmt, err := sqlair.Prepare(getUpgradeDBCompletedQuery, Info{}, sqlair.M{})
+	if err != nil {
+		return errors.Annotatef(err, "preparing %q", getUpgradeDBCompletedQuery)
+	}
+
+	completeDBUpgradeQuery := "UPDATE upgrade_info SET db_completed_at = DATETIME('now') WHERE uuid = $M.info_uuid"
```
Comment (manadart) [inline domain/upgrade/state/state.go:315]: I would not do a separate select before attempting to update.
- Add `where db_completed_at IS NULL` to the update statement.
- Check the `RowsAffected` result from running the update.
- If there are no rows affected, return an error saying that there is no upgrade record with a pending DB upgrade.

worker/changestreampruner/worker.go has an example of getting a `sql.Outcome` and checking the result.

**Review by manadart — COMMENTED** [review]
We shouldn't blithely accept that a DB upgrade has already been completed when the worker calls this new method. It's an indicator that something is wrong and our wires are crossed.

File: domain/upgrade/state/types.go [hunk 3]
```diff
@@ -5,16 +5,28 @@ package state
...
 import "database/sql"
 
-type info struct {
-	UUID            string         `db:"uuid"`
-	PreviousVersion string         `db:"previous_version"`
-	TargetVersion   string         `db:"target_version"`
-	CreatedAt       string         `db:"created_at"`
-	StartedAt       sql.NullString `db:"started_at"`
-	CompletedAt     sql.NullString `db:"completed_at"`
+// These structs represent the upgrade entity schema in the database.
```
Comment (manadart) [inline domain/upgrade/state/types.go:8]: This comment doesn't add anything. Better to comment the actual types.

File: domain/upgrade/state/types.go [hunk 4]
```diff
@@ -5,16 +5,28 @@ package state
...
+
+type Info struct {
```
Comment (manadart) [inline domain/upgrade/state/types.go:10]: These don't need to be exported. They are not used outside of this package.

File: domain/upgrade/state/state.go [hunk 5]
```diff
@@ -299,3 +299,33 @@ func (st *State) ActiveUpgrade(ctx context.Context) (string, error) {
...
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var outcome sqlair.Outcome
+		if err = tx.Query(ctx, completedDBUpgradeStmt, sqlair.M{"info_uuid": upgradeUUID}).Get(&outcome); err != nil {
+			return errors.Trace(err)
+		}
+		if num, err := outcome.Result().RowsAffected(); err != nil {
+			return errors.Trace(err)
+		} else if num == 0 {
+			return errors.Errorf("There is no upgrade record with a pending DB upgrade ID: %s", upgradeUUID)
```
Comment (manadart) [inline domain/upgrade/state/state.go:327]: Errors generally don't start with capitals.
This should be: 
```go
return errors.NotFoundf("current upgrade with ID %q and pending database update")
```

File: domain/upgrade/service/service.go [hunk 6]
```diff
@@ -85,6 +86,15 @@ func (s *Service) SetControllerDone(ctx context.Context, upgradeUUID, controller
 	return domain.CoerceError(s.st.SetControllerDone(ctx, upgradeUUID, controllerID))
 }
 
+// SetDBUpgradeCompleted marks the upgrade as completed in the database
+func (s *Service) SetDBUpgradeCompleted(ctx context.Context, upgradeUUID string) error {
+	err := s.st.SetDBUpgradeCompleted(ctx, upgradeUUID)
+	if database.IsErrNotFound(err) {
```
Comment (manadart) [inline domain/upgrade/service/service.go:92]: You can't get this error now, so remove this case.

**Review by manadart — COMMENTED** [review]
Almost there.

File: domain/upgrade/state/state.go [hunk 7]
```diff
@@ -299,3 +299,33 @@ func (st *State) ActiveUpgrade(ctx context.Context) (string, error) {
 	})
 	return activeUpgrade, errors.Trace(err)
 }
+
+func (st State) SetDBUpgradeCompleted(ctx context.Context, upgradeUUID string) error {
```
Comment (manadart) [inline domain/upgrade/state/state.go:303]: You need a test for this method.

Comment (anvial) [reply domain/upgrade/state/state.go:303]: thx, done

File: domain/upgrade/state/state_test.go [hunk 8]
```diff
@@ -326,3 +326,13 @@ func (s *stateSuite) TestActiveUpgradesSingular(c *gc.C) {
 	c.Assert(err, jc.ErrorIsNil)
 	c.Check(activeUpgrade, gc.Equals, uuid)
 }
+
+func (s *stateSuite) TestSetDBUpgradeCompleted(c *gc.C) {
+	uuid, err := s.st.CreateUpgrade(context.Background(), version.MustParse("3.0.0"), version.MustParse("3.0.1"))
+	c.Assert(err, jc.ErrorIsNil)
+	err = s.st.SetDBUpgradeCompleted(context.Background(), uuid)
+	c.Assert(err, jc.ErrorIsNil)
+
```
Comment (manadart) [inline domain/upgrade/state/state_test.go:335]: You could call it again and assert that you get a `NotFound` error, to test the negative case.

**Review by manadart — APPROVED** [review]
Thanks for the changes.

File: domain/upgrade/state/state_test.go (hunk 8 above)
Comment (anvial) [reply domain/upgrade/state/state_test.go:338]: done

**Pull request merged**